File: 20220918-N_Queen-Java/app/src/main/java/NQueen.java

// Q 0 0 0  [0] [1] [2] [3]
// 0 0 0 0  [0] [1] [2] [3]
// 0 0 0 0  [0] [1] [2] [3]
// 0 0 0 0  [0] [1] [2] [3]

import java.util.*;

public class NQueen {
  public int solution(int n) {
    int[] column;
    int answer = 0;

    column = new int[n];
    for (int i = 0; i < n; i += 1) {
      column[i] = -1;
    }

    answer = dfs(n, 0, column);
    return answer;

  }

  public int dfs(int n, int index, int[] column) {
    if (index == n) {
      return 1;
    }
    int res = 0;
    for (int i = 0; i < n; i++) {
      column[index] = i;
      if (check(index, column)) res += dfs(n, index + 1, column);
      column[index] = -1;
    }
    return res;
  }

  public boolean check(int here, int[] column) {
    for (int i = 0; i < here; i++) {
      if (column[here] == column[i]) return false;
      if (Math.abs(column[here] - column[i]) == here - i) return false;
    }
    return true;
  }
}

File: 20220920-NQueen-javaScript/NQueen.js
function solution(n) {
  let answer = 0;

  const isValid = (board, row) => {
    for (let i = 1; i < row; i += 1) {
      if (board[i] === board[row]) return false;
      if (Math.abs(board[i] - board[row]) === Math.abs(i - row)) return false;
    }
    return true;
  };

  const dfs = (board, row) => {
    if (row === n) answer += 1;
    else {
      for (let i = 1; i <= n; i += 1) {
        if (isValid(board, row + 1)) {
          dfs(board, row + 1);
        }
      }
    }

    for (let i = 1; i <= n; i += 1) {
      const board = new Array(n + 1).fill(0);
      board[1] = i;
      dfs(board, 1);
    }
  };

  return answer;
}
